x/authz/count_authorization.go

```diff
@@ -0,0 +1,49 @@
+package authz
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+var (
+	_ Authorization = &CountAuthorization{}
+)
+
+// NewCountAuthorization creates a new CountAuthorization object.
+func NewCountAuthorization(msgTypeURL string, allowedAuthorizations int32) *CountAuthorization {
+	return &CountAuthorization{
+		Msg:                   msgTypeURL,
+		AllowedAuthorizations: allowedAuthorizations,
+	}
+}
+
+// MsgTypeURL implements Authorization.MsgTypeURL.
+func (a CountAuthorization) MsgTypeURL() string {
+	return a.Msg
+}
+
+// Accept implements Authorization.Accept.
+func (a CountAuthorization) Accept(ctx sdk.Context, msg sdk.Msg) (AcceptResponse, error) {
+	remaining, isNegative := a.decrement()
+	if isNegative {
+		return AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrapf("allowed authorizations cannot be negative")
+	}
+	if remaining == 0 {
+		return AcceptResponse{Accept: true, Delete: true}, nil
+	}
+
+	return AcceptResponse{Accept: true, Delete: false, Updated: &CountAuthorization{Msg: a.Msg, AllowedAuthorizations: remaining}}, nil
+}
+
+// ValidateBasic implements Authorization.ValidateBasic.
+func (a CountAuthorization) ValidateBasic() error {
+	if a.AllowedAuthorizations < 0 {
+		return sdkerrors.ErrInvalidRequest.Wrapf("allowed authorizations cannot be negative")
+	}
+	return nil
```

Comment:
An `AllowedAuthorizations` of 0 should count as invalid since it means there are no uses available. So it should be `<= 0` and the message should be changed to something like "allowed authorizations must be greater than 0."